Apple Pay & Google Pay buttons no longer visible in Standard Payments button previews after moving them to Advanced Card Processing tab (3179)

stracker-phil opened the pull request:
## Problem
After moving Google Pay and Apple Pay settings to the "**Advanced Card Processing**" tab, the button preview was not available anymore - neither in the "Standard Payments" nor the "Advanced Card Processing" tab.

## Background
1. Advanced Card Processing: Did not render the preview because relevant JS files were not enqueued
2. Standard Payments: Did not render the preview because it was tightly coupled to the presence of certain DOM elements (eg the presence of an "is-enabled" checkbox)

## Solution
1. Refactor preview button code to decouple it from DOM elements
2. Enqueue relevant JS files in both tabs

## Changes

### Preview button logic

- Adjust the preview button generator to render an individual button (before this, the button preview always showed a block of all PayPal payment buttons)
- Layout of the single preview button should be as close to other "Button Styling Preview" blocks as possible
- On the "Standard Payments" tab, the APM buttons should be static and not change colors/labels when changing any settings

### Overview of code changes

#### `ppcp-button/.../PreviewButton.js` 
- Contains the common logic to render a single APM button preview
- "Abstract" base class, some methods must be implemented in APM module

#### `ppcp-button/.../PreviewButtonManager.js`
- Extracted common logic to manage _one or multiple_ APM `PreviewButton`s
- Bootstrap logic that loads remote configuration from PayPal
- "Abstract" base class, some methods must be implemented in APM module

#### Both `boot-admin.js`
- Implement child classes of `PreviewButton` and `PreviewButtonManager`

#### `ppcp-wc-gateway/services.php`
- Condition now loads the APM's `boot-admin.js` in both tabs ("Standard Payments" and "Advanced Card Processing")

#### Both `extensions.php`
- APM edit form includes a button preview (at the end)
- Custom attributes for edit-fields, to trigger an update of the button preview

#### `gateway-settings.js`
- New light-weight form-observer that interacts with the changes in `extensions.php`.
- The new observer code only renders & refreshes a single APM button

### Screenshots
Fixed preview of the full button block, in the "Standard Payments" tab:
![sp-button-block-preview](https://github.com/woocommerce/woocommerce-paypal-payments/assets/2669200/0cefc9cb-ba27-40f8-9ab2-75c97f9d714c)

New preview of the single payment button, in the "Advanced Card Processing" tab:
![acp-single-button-preview](https://github.com/woocommerce/woocommerce-paypal-payments/assets/2669200/f06a2f81-0531-47e5-8bf2-aa4f5ccb5e91)

stracker-phil committed:
🎨 Isolate condition for preview buttons

This step makes it easier to refactor/update the logic in the next step.

stracker-phil committed:
♻️ Refactor Google Pay preview button

Mainly, separate the preview code into three main parts:
- Configuration
- Button
- ButtonManager

Remove dependency from DOM structure

stracker-phil committed:
♻️ Create generic base classes for preview buttons

stracker-phil committed:
🐛 Load button preview JS on ACDC settings page

Renamed the DI service name; this service is only used in two places (APM modules), which were also updated.

stracker-phil committed:
🔥 Remove debug code

stracker-phil committed:
🔀 Merge branch 'trunk' into PCP-3179-apple-pay-google-pay-buttons-no-longer-visible-in-standard-payments-button-previews-after-moving-them-to-advanced-card-processing-tab

stracker-phil committed:
♻️ Fix logic in preview button base classes

- Promise-based initialization is executed in correct order
- Support to redraw single APM type of button
- Fix the button style
- Simplify the PreviewButton code

stracker-phil committed:
♻️ Clean up the Google Pay preview button code

stracker-phil committed:
🔥 Remove old preview button code comment

stracker-phil committed:
✨ Add Google Pay preview button to ACP tab

Note: `phpcbf` adjusted tons of whitespace.

stracker-phil committed:
🔀 Merge branch 'trunk'

stracker-phil committed:
✨ Add preview button for Apple Pay on ACDC

Note: `phpcbf` adjusted whitespace in extensions.php

stracker-phil committed:
🐛 Fix button specific config-preparation

stracker-phil committed:
🦺 Add check for config that causes infinite loop

stracker-phil committed:
✨ Preview buttons can be static

When the current page does not contain any edit-fields for the current APM preview button, it enters “static” mode. Static buttons ignore most update requests, and render as they were defined during page load

stracker-phil committed:
🔀 Merge branch 'trunk'

stracker-phil committed:
🔥 Remove old preview button code comment

stracker-phil committed:
🔥 Remove unused yarn package

stracker-phil committed:
🔥 Remove unused import

Dinamiko commented on modules/ppcp-button/resources/js/modules/Renderer/PreviewButton.js:
```diff
@@ -0,0 +1,135 @@
+     * @return {this} Reference to self, for chaining.
+     */
+    setPpcpConfig(config) {
+        this.ppcpConfig = merge({}, config);
```
Probably I'm missing something here... but why merge an empty object here? how is this different from:
```
this.ppcpConfig = config;
```

stracker-phil replied:
Good catch 👍 The idea of this part is, to create a local copy of the config object.

Since the config object can be modified outside the class, we want to keep a static version of the provided configuration.

```js
this.button[id].setPpcpConfig(config)
...
config.color = 'black'
```

Without the "merge", this would cause some unexpected behavior; with "merge" changing the config object outside the class does not impact the preview, until explicitly calling `setPpcpConfig()` again.

Dinamiko commented on modules/ppcp-button/resources/js/modules/Renderer/PreviewButton.js:
```diff
@@ -0,0 +1,135 @@
+
+if (buttonWrapper === ppcpWrapper) {
+throw new Error(`[APM Preview Button] Infinite loop detected. Provide different selectors for the button/ppcp wrapper elements! Selector: "#${buttonWrapper}"`);
+}
```
We do not have the linter installed yet... but just indent this a bit for now.

Dinamiko commented on modules/ppcp-button/resources/js/modules/Renderer/PreviewButtonManager.js:
```diff
@@ -0,0 +1,218 @@
+     * @param {string} wrapperId - CSS ID of the wrapper element.
+     * @return {PreviewButton}
+     */
+    createButtonInst(wrapperId) {
```
Better use complete words: `createButtonInstance`.

stracker-phil committed:
🐛 Improve error logging and dependency detection

stracker-phil committed:
🔀 Merge branch 'trunk'

stracker-phil committed:
🔀 Merge branch 'trunk'

stracker-phil committed:
♻️ Improve SDK loading in PreviewButtonManager

Dinamiko reviewed (approved):
It looks good, PR is approved, I've added a couple of minor adjustments.

Regarding jQuery I’ve noticed that most of the code was already there before you started the implementation, we'll leave it there for now but just for you to know that the idea would be to progressively refactor to vanilla JS, so not in this PR but from now on each opportunity found to refactor it, just do it.

stracker-phil committed:
🎨 Minor code structure changes

stracker-phil committed:
🦺 Add new console error to help us debug issues

This new console error should help us identify cache-related issues faster

stracker-phil committed:
🎨 Minor: Apply WP Code Style

- Consistent quote-style
- Insert missing semicolons in JS
- Fix some whitespaces

stracker-phil committed:
🐛 Hide disabled APM buttons in Standard Payments

stracker-phil committed:
🎨 Remove unnecessary condition

stracker-phil committed:
🐛 Render preview when enabling the button

Before this, the “is enabled” checkbox was not observed; enabling it did not cause a preview button to render, and required changing any other form field that was linked to the button

Also start to extract some logic to a new helper file.

stracker-phil committed:
🐛 Fix the condition to inject the settings JS

Switched to using an existing DI service to determine if the current page requires the “gateway-settings” asset files.

stracker-phil committed:
🐛 Refresh all preview buttons on the screen

When multiple preview-button-blocks for the same button were present on the page, not all of them did update the button.

stracker-phil committed:
🩹 Fix incorrect usage of #-prefix

Only properties can be declared as private using `#`. For functions, this is not supported yet, using the private-prefix here is unreliable.

stracker-phil committed:
🐛 Fix hidden preview buttons in Standard Payments

Addresses an issue when toggling the “Customize per location” checkbox multiple times.

stracker-phil committed:
🐛 Force-display the preview button after render()

Some tests indicate, that the `ppcp-hidden` and `ppcp-enabled` events are not always triggered on some sites. As the preview button is refreshed during those events, it would stay hidden on those sites.

Adding a delayed DOM show() call solves the issue in a slightly less-elegant way.
